pkg/block/fetcher.go

```diff
@@ -574,4 +596,37 @@ func (f *ConsistencyDelayMetaFilter) Filter(metas map[ulid.ULID]*metadata.Meta,
 			delete(metas, id)
 		}
 	}
+
+	return nil
+}
+
+// DeletionMarkMetaFilter is a MetaFetcher that filters blocks that are marked for deletion (deleteDelay/2) duration before current time.
+type DeletionMarkMetaFilter struct {
+	logger      log.Logger
+	deleteDelay time.Duration
+	bkt         objstore.BucketReader
+}
+
+// NewDeletionMarkMetaFilter creates DeletionMarkMetaFilter.
+func NewDeletionMarkMetaFilter(logger log.Logger, bkt objstore.BucketReader, deleteDelay time.Duration) *DeletionMarkMetaFilter {
+	return &DeletionMarkMetaFilter{
+		logger:      logger,
+		bkt:         bkt,
+		deleteDelay: deleteDelay,
+	}
+}
+
+// Filter filters out blocks that are marked for deletion (deleteDelay/2) duration before current time.
+func (f *DeletionMarkMetaFilter) Filter(ctx context.Context, metas map[ulid.ULID]*metadata.Meta, synced GaugeLabeled, _ bool) error {
+	for id := range metas {
+		deletionMark, err := metadata.ReadDeletionMark(ctx, f.bkt, f.logger, id.String())
+		if err != nil && errors.Cause(err) != metadata.ErrorDeletionMarkNotFound {
+			return errors.Wrapf(err, "read file deletion-mark file for block: %s", id.String())
+		}
+		if deletionMark != nil && time.Since(time.Unix(deletionMark.DeletionTime, 0)).Seconds() > f.deleteDelay.Seconds()/2 {
```

Comment:
Prefer easy path here. Also, avoid using var that is returned with err != nil. What about:

```
if err == metadata.ErrorDeletionMarkNotFound {
  continue
}
```